test/functional/p2p_v2_misbehaving.py

```diff
@@ -0,0 +1,201 @@
+#!/usr/bin/env python3
+# Copyright (c) 2022 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+import random
+from enum import Enum
+
+from test_framework.crypto.ellswift import ellswift_create
+from test_framework.p2p import P2PInterface
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import random_bitflip
+from test_framework.v2_p2p import (
+    logger,
+    EncryptedP2PState,
+    MAX_GARBAGE_LEN,
+    TRANSPORT_VERSION,
+)
+
+
+class TestType(Enum):
+    """ Scenarios to be tested:
+
+    1. EARLY_KEY_RESPONSE - The responder needs to wait until one byte is received which does not match the 16 bytes
+    consisting of network magic followed by "version\x00\x00\x00\x00\x00" before sending out it's ellswift + garbage bytes
+    2. EXCESS_GARBAGE - Disconnection happens when > MAX_GARBAGE_LEN bytes garbage is sent
+    3. WRONG_GARBAGE_TERMINATOR - Disconnection happens when incorrect garbage terminator is sent
+    4. WRONG_GARBAGE - Disconnection happens when garbage bytes that is sent is different from what the peer receives
+    5. SEND_NO_AAD - Disconnection happens when AAD of first encrypted packet after the garbage terminator is not filled
+    6. SEND_NON_EMPTY_VERSION_PACKET - non-empty version packet is simply ignored
+    """
+    EARLY_KEY_RESPONSE = 0
+    EXCESS_GARBAGE = 1
+    WRONG_GARBAGE_TERMINATOR = 2
+    WRONG_GARBAGE = 3
+    SEND_NO_AAD = 4
+    SEND_NON_EMPTY_VERSION_PACKET = 5
+
+
+class TestEncryptedP2PState(EncryptedP2PState):
+    """ Modify v2 P2P protocol functions for testing scenarios listed in `TestType`"""
+    def __init__(self, test_type):
+        super().__init__(initiating=True, net='regtest')
+        self.test_type = test_type
+        if test_type == TestType.EARLY_KEY_RESPONSE:
+            self.magic_sent = False  # set to True after first 4 bytes of ellswift which match network magic is sent.
+            self.can_data_be_received = False  # variable used to assert if data is received on recvbuf.
+
+    def generate_keypair_and_garbage(self):
+        """Generate > MAX_GARBAGE_LEN garbage bytes, MAX_GARBAGE_LEN//2 garbage bytes, tampered garbage bytes
+        when TestType = (EXCESS_GARBAGE, WRONG_GARBAGE_TERMINATOR, WRONG_GARBAGE)"""
+        self.privkey_ours, self.ellswift_ours = ellswift_create()
+
+        if self.test_type == TestType.EXCESS_GARBAGE:
+            # send > 4095 bytes garbage
+            garbage_len = MAX_GARBAGE_LEN + random.randrange(1, 10)
+        elif self.test_type == TestType.WRONG_GARBAGE_TERMINATOR:
+            garbage_len = random.randrange(MAX_GARBAGE_LEN//2)
+        else:
+            garbage_len = random.randrange(MAX_GARBAGE_LEN + 1)
```

Comment:
doesn't the `+1` make the `garbage_len` greater than `4095` which is what is being covered in `TestType.EXCESS_GARBAGE`

Reply:
never mind randrange doesn't include the value passed

Reply:
yes!